features/lorawan/lorastack/phy/LoRaPHY.cpp

```diff
@@ -123,316 +106,1317 @@ void LoRaPHY::handle_send(uint8_t *buf, uint8_t size)
+    uint32_t frequency = rx_conf->frequency;
+
+    _radio->lock();
+
+    if (_radio->get_status() != RF_IDLE) {
+        _radio->unlock();
+        return false;
+    }
+
+    _radio->unlock();
+
+    if (rx_conf->rx_slot == RX_SLOT_WIN_1) {
+        // Apply window 1 frequency
+        frequency = phy_params.channels.channel_list[rx_conf->channel].frequency;
+        // Apply the alternative RX 1 window frequency, if it is available
+        if (phy_params.channels.channel_list[rx_conf->channel].rx1_frequency != 0) {
+            frequency = phy_params.channels.channel_list[rx_conf->channel].rx1_frequency;
+        }
+    }
+
+    // Read the physical datarate from the datarates table
+    uint8_t *datarate_table = (uint8_t *) phy_params.datarates.table;
+    uint8_t *payload_table = (uint8_t *) phy_params.payloads.table;
+    uint8_t *payload_with_repeater_table = (uint8_t *) phy_params.payloads_with_repeater.table;
+
+    phy_dr = datarate_table[dr];
+
+    _radio->lock();
+
+    _radio->set_channel(frequency);
+
+    // Radio configuration
+    if (phy_dr == DR_7 && phy_params.fsk_supported) {
+        modem = MODEM_FSK;
+        _radio->set_rx_config(modem, 50000, phy_dr * 1000, 0, 83333, 5,
+                              rx_conf->window_timeout, false, 0, true, 0, 0,
+                              false, rx_conf->is_rx_continuous);
+    } else {
+        modem = MODEM_LORA;
+        _radio->set_rx_config(modem, rx_conf->bandwidth, phy_dr, 1, 0, 8,
+                              rx_conf->window_timeout, false, 0, false, 0, 0,
+                              true, rx_conf->is_rx_continuous);
+    }
+
+    _radio->unlock();
+
+    if (rx_conf->is_repeater_supported) {
+        max_payload = payload_with_repeater_table[dr];
+    } else {
+        max_payload = payload_table[dr];
+    }
+
+    _radio->lock();
+
+    _radio->set_max_payload_length(modem, max_payload + LORA_MAC_FRMPAYLOAD_OVERHEAD);
+
+    _radio->unlock();
+
+    *datarate = phy_dr;
+
+    return true;
+}
+
+bool LoRaPHY::tx_config(tx_config_params_t* tx_conf, int8_t* tx_power,
+                        lorawan_time_t* tx_toa)
+{
+    radio_modems_t modem;
+    int8_t phy_dr = ((uint8_t *)phy_params.datarates.table)[tx_conf->datarate];
+    channel_params_t *list = phy_params.channels.channel_list;
+    uint8_t band_idx = list[tx_conf->channel].band;
+    band_t *bands = (band_t *)phy_params.bands.table;
+
+    // limit TX power if set to too much
+    if (tx_conf->tx_power > bands[band_idx].max_tx_pwr) {
+        tx_conf->tx_power = bands[band_idx].max_tx_pwr;
+    }
+
+    uint8_t bandwidth = get_bandwidth(tx_conf->datarate);
+    int8_t phy_tx_power = 0;
+
+    // Calculate physical TX power
+    phy_tx_power = compute_tx_power(tx_conf->tx_power, tx_conf->max_eirp,
+                                    tx_conf->antenna_gain);
+
+    // Setup the radio frequency
+    _radio->lock();
+    _radio->set_channel(list[tx_conf->channel].frequency);
+    _radio->unlock();
```

Comment:
I don't think we need lock/unlock separately for each radio access. For example in this method we call radio if several times so I think we need only one lock/unlock.

Reply:
Same comment applies for some other methods in this class too.

Reply:
I was trying to be a bit fine grained here, but may be its not that critical.

Reply:
if there is no any long lasting operation between radio call's then one lock is enough.